opentelemetry-api/src/opentelemetry/metrics/__init__.py

```diff
@@ -11,3 +11,513 @@
+
+    :class:`Measurement`s are used for recording raw values, in which the
+    aggregation and labels for the exported metric are defered. This should be
+    used to record measurements like "server_latency" or "received_bytes",
+    where the value of interest is the recorded value itself.
+
+    :class:`Metric`s are used for recording pre-defined aggregation, or already
+    aggregated data. This should be used to report metrics like cpu/memory
+    usage, in which the type of aggregation is already defined, or simple
+    metrics like "queue_length".
+
+    Initialization of the :class:`Meter` is done through the `meter()` function
+    (see below). The meter instance has a singleton implementation.
+    """
+
+    def create_double_counter(self,
+                              name: str,
+                              options: typing.Optional['MetricOptions'] = None
+                              ) -> 'CounterDouble':
+        """Creates a counter type metric that contains double values.
+
+        Args:
+            name: The name of the counter.
+            options: An optional :class:`.MetricOptions` used to populate
+            details about the counter.
+
+        Returns: A new :class:`.CounterDouble`
+        """
+
+    def create_long_counter(self,
+                            name: str,
+                            options: typing.Optional['MetricOptions'] = None
+                            ) -> 'CounterLong':
+        """Creates a counter type metric that contains long values.
+
+        Args:
+            name: The name of the counter.
+            options: An optional :class:`.MetricOptions` used to populate
+            details about the counter.
+
+        Returns:
+            A new :class:`.CounterLong`
+        """
+
+    def create_double_gauge(self,
+                            name: str,
+                            options: typing.Optional['MetricOptions'] = None
+                            ) -> 'GaugeDouble':
+        """Creates a gauge type metric that contains double values.
+
+        Args:
+            name: The name of the gauge.
+            options: An optional :class:`.MetricOptions` used to populate
+            details about the gauge.
+
+        Returns:
+            A new :class:`.GaugeDouble`
+        """
+
+    def create_long_gauge(self,
+                          name: str,
+                          options: typing.Optional['MetricOptions'] = None
+                          ) -> 'GaugeLong':
+        """Creates a gauge type metric that contains long values.
+
+        Args:
+            name: The name of the gauge.
+            options: An optional :class:`.MetricOptions` used to populate
+            details about the gauge.
+
+        Returns:
+            A new :class:`.GaugeLong`
+        """
+
+    def create_measure(self,
+                       name: str,
+                       options: typing.Optional['MeasureOptions'] = None
```

Comment:
According to the [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-metrics.md#createdoublemeasurement), a measurement has both `createDouble` and `createLong` methods, irregardless of the type of the `measure` that was used to create the `measurement`. If we remove the type in `MeasureOptions`,  there will be no validation logic between the `measure_type` and the type of `measurement` being created. I think we want to prevent a `measure` to be able to create any type of `measurement`.

Reply:
This looks to me like the spec was written to match the java client, and as far as I can tell there's no reason for (java's) `Measure` to have both `createDoubleMeasurement` and `createLongMeasurement`.

I'd imagine something like this instead:
```python
class Meter:
    def create_double_measure(...)
    def create_long_measure(...)

class DoubleMeasure(Measure):
    def create_measurement(value)
        # returns a double-valued Measurement

class LongMeasure(Measure):
    def create_measurement(value)
        # returns a long-valued Measurement
```

and possibly replacing `Measurement` with separate `DoubleMeasurement` and `LongMeasurement` classes.

I don't mean to suggest deviating from the spec. The spec is underdefined here and writing the python client should help us generate spec changes.

Reply:
You're right. I was focused too much on following the spec. I also think the ability to create both types of measurements and having to validate the type of measure is redundant as well. I will make these changes and propose them in the spec.